Batching rule for Tensor.new_empty_strided

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47227 Batched gradient support for view+inplace operations
* #47365 Add batching rule for torch.clone(tensor, torch.contiguous_format)
* #47364 as_strided batching rule
* **#47226 Batching rule for Tensor.new_empty_strided**
* #47225 Implement Tensor.new_empty_strided(sizes, strides, *, dtype, device, requires_grad)

The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

Differential Revision: [D24741690](https://our.internmc.facebook.com/intern/diff/D24741690)

zou3519 committed:
Batching rule for Tensor.new_empty_strided

The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
+  // - physical size: [B0, B1, B2, 5, 3]
+  // - physical stride: [100*B1*B2, 100*B2, 100, 0, 1]
+  //
+  // Let's select the strides that result in the smallest memory usage.
```
I am not sure about this argument. Even we we transposed everything to be "column contiguous" the actual memory usage would be the same no?
Is the choice here done to make the batch tensor contiguous if the original result of `empty_strided(size, stride)` was contiguous?

zou3519 replied:
If we picked `[100*B1*B2, 100*B2, 100, 0, 1]` instead of `[B1 * B2 *S, B2 * S, S, 5, 3]`, then the resulting storage size would be very large (100*B1*B2 * (B0-1) would be the memory location of result[B0-1], rather than B1 * B2 *S * (B0-1))

If the original result of `empty_strided(size, stride)` is contiguous, then the batched tensor is contiguous, yes -- that is also one of the motivations (I'll add that to the comment)

What do you mean by "column contiguous"?

albanD replied:
> What do you mean by "column contiguous"?

Sorry that is not very clear. Stride being strictly increasing. Matching a column-major packing.

> If we picked [100*B1*B2, 100*B2, 100, 0, 1] instead of [B1 * B2 *S, B2 * S, S, 5, 3]

You mean using another fixed values V (=100 here) such that S <= V?
I guess that is true but I'm not sure why would anyone argue that this is a thing we want to do?

zou3519 replied:
> You mean using another fixed values V (=100 here) such that S <= V?
I guess that is true but I'm not sure why would anyone argue that this is a thing we want to do?

Yes, exactly. It's a degree of freedom that I was shutting down for completeness. I'll refactor the comment to make it clear that the reason is more to make the batch tensor contiguous if the original result of empty_strided(size, stride) is contiguous (and if the original result is not contiguous, for the batch dimensions to be contiguous in memory)

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
+  auto physical_view = MultiBatchVmapTransform::logicalToPhysical(self);
+  auto physical_size = physical_view.getPhysicalShape(size);
+
+  // Let [B0, B1, B2] be the shape of the batch dims. We know what the physical
```
What happens if all the batch dimensions are not at the front?

zou3519 replied:
MultiBatchVmapTransform::logicalToPhysical(self) moved all of the batch dims to the front

zou3519 replied:
But I should clarify in the comment that we are going to create the batch dimensions at the front, even if in the original tensors being vmapped the batch dim wasn't at the front.

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
+        "new_empty_strided(sizes, strides): dimensionality of sizes (",
+        size.size(), ") must match dimensionality of strides (",
+        stride.size(), ")");
+  auto example_storage_size = native::storage_size_for(size, stride);
```
nit: why is this one called "example"?

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -694,6 +695,53 @@ Tensor new_empty_batching_rule(
+  // physical_strides = [B1 * B2 * S, B2 * S, S] + strides
+  physical_strides.insert(physical_strides.end(), stride.begin(), stride.end());
+
+  auto result = physical_view.tensor().new_empty_strided(
```
nit: Do you have to call back into `new_empty_strided` here? Or could it be done with `at::empty_strided` and this is just for consistency?

zou3519 replied:
This can be done with at::empty_strided. I chose to use `new_empty_strided` because it was the first thing that came to mind. I like at::empty_strided because it's less verbose, let's go with that.

zou3519 replied:
Actually, I take that back: there is indeed a semantic difference. If the user didn't provide some of the options, then we use the options of the original tensor. For example:

```
x = torch.rand(3, device='cuda')
y = x.new_empty_strided([4, 4], [4, 1])
```
this should create a new CUDA tensor. However, none of the options being passed to `new_empty_strided_batching_rule` have actually been set. So if we used `at::empty(physical_size, physical_strides, options)`, this would end up creating a CPU tensor (the default for device when it is not set is CPU)

albanD commented on aten/src/ATen/native/Resize.h:
```diff
@@ -51,15 +52,7 @@ inline TensorImpl* resize_impl_cpu_(
     self->set_sizes_and_strides(size, *stride);
     // NB: storage size can be different from numel.
-    for (size_t dim = 0; dim < size.size(); ++dim) {
-      // FIXME: Don't rely on storage_size being negative because this
```
I see that this comment was dropped. Do you know what it was referring to?

zou3519 replied:
I wrote this comment two years ago but unfortunately I don't remember why :/

Looking at [the original PR](https://github.com/pytorch/pytorch/pull/13185/files#diff-67630b8526b429e0fc2c7851f15eb290808c90d898e8ab98fb28821d513c3a46R38), previously, storage_size could be negative if size[dim] is equal to 0. I think that is what the comment is referring to.

However, in the current iteration of the code, storage_size can never be negative, so I think the comment doesn't apply anymore.

albanD replied:
Right the dim=0 case is special cased so that should be fine!

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

albanD reviewed (approved):
lgtm

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

[ghstack-poisoned]

zou3519 committed:
Update on "Batching rule for Tensor.new_empty_strided"


The batching rule is a little weird because it's not immediately obvious
what the strides of the result should be. If
tensor.new_empty_strided(size, stride) is called inside vmap and
`tensor` is being vmapped over, the result is a physical tensor with:
- size `[batch_shape] + size`
- strides `[S0, S1, ..., Sn] + stride` such that the
S0...Sn are part of a contiguous subspace and Sn is equal to the size of
the storage of `torch.empty_strided(size, stride)`.

I refactored some of the logic that computes the storage size for
`torch.empty_strided(size, stride)` into a helper function
`native::storage_size_for` and use it in the batching rule.

Test Plan:
- New tests in test/test_vmap.py

Differential Revision: [D24741690](https://our.internmc.facebook.com/intern/diff/D24741690)

[ghstack-poisoned]
